### PR TITLE
feat(events): granular event visibility settings — hide capacity, counts, and guest list

### LINKED ISSUE #792: feat(events): granular event visibility settings — hide capacity, counts, and guest list
## Motivation

Organizer request (via rep): run an event with fully hidden organizational information — number of slots before the waiting list, already-booked slots, and no guest names even if guests individually allow visibility.

Today this is not possible: `attendee_count`, `max_attendees`, and `seats_held` live on `EventBaseSchema` (`src/events/schema/event.py`) and are served **to anonymous users** on every list/calendar/detail endpoint with zero gating. The attendee list is already per-user opt-in (`show_me_on_attendee_list`, default `NEVER`), but there is no event-level override.

## Design (agreed)

Add a `visibility_settings` JSONField on `Event`, validated by a pydantic `EventVisibilitySettings` model (typed in OpenAPI so the FE gets a real object). Follows the existing `Event.schedule` JSONField-with-validator pattern.

| Toggle | Type | Default (= current behavior) |
|---|---|---|
| `show_attendee_count` | bool | `true` |
| `show_capacity` — gates `max_attendees`, `seats_held`, derived spots-left, **and** `TicketTierSchema.total_available` | bool | `true` |
| `show_attendee_list` | bool | `true` — **ANDed** with per-user `show_me_on_attendee_list` prefs (can only hide, never reveal) |

Additional decisions:

- **`is_full` derived boolean is always public**, even when counts are hidden — FE needs it for sold-out badges and to explain blocked RSVPs. "Full yes/no" is not sensitive; exact numbers are.
- **Enforcement pattern**: context-aware schema resolvers backed by cached model predicates, per the existing `resolve_address` → `Event.can_user_see_address(user)` idiom. Staff/owners always see everything. Do **not** replicate the controller-403 or service-filter patterns.
- `resolve_seats_held` currently takes no `context` argument — needs a signature change to gate.
- Data migration is trivial: defaults mirror today's behavior, existing events change nothing. Preset bundles ("discreet event" etc.) are FE-only UX; backend stores granular values.
- New Event fields are copied to recurring occurrences by default (see `events.service.duplication._EXCLUDED_FROM_COPY`) — copying `visibility_settings` to occurrences is the desired behavior, so no exclusion needed. Template editors must round-trip the field.

## Full leak surface to gate (checklist)

Hiding only the headline count is trivially bypassed — all of these expose counts:

- [ ] `EventBaseSchema.attendee_count` (anonymous, list + detail + calendar)
- [ ] `EventBaseSchema.max_attendees` (anonymous)
- [ ] `EventBaseSchema.resolve_seats_held` (anonymous; no `context` today)
- [ ] `TicketTierSchema.total_available` via `GET /events/{id}/tickets/tiers` (anonymous)
- [ ] `EventPronounDistributionSchema.total_attendees` / `total_with_pronouns` / `total_without_pronouns` — a second exact attendee count behind a *different* toggle (`public_pronoun_distribution`); when counts are hidden, this endpoint must not leak totals to non-staff
- [ ] Dietary summary per-restriction `attendee_count` (authenticated) — gate aggregate counts for non-staff when counts are hidden
- [ ] Attendee-list endpoint: apply `show_attendee_list` AND before the `AttendeeVisibilityFlag` matrix

Not leaking today (verified, no action): total waitlist size is never public; waitlist position / `pending_offers_count` are self-only via `/my-status`.

## Known ceiling (document, don't fix)

Seated ticketed events inherently reveal occupancy through the seat picker (`GET .../seating/availability` returns a per-seat sold/held map — anonymous). The seat map cannot both work and hide occupancy. Scope: `show_capacity` gates tier `total_available` numbers, but the seat map stays as-is; document the limitation for organizers.

## Out of scope

- Migrating `address_visibility` and `public_pronoun_distribution` into `visibility_settings` — phase 2 follow-up (see linked issue), breaking OpenAPI contract change, droppable if it never earns its cost.
- Backend-side visibility presets (YAGNI — FE concern).

## Testing

- Wire-contract tests: anonymous vs authenticated vs attendee vs staff for each gated field, toggles on/off.
- `is_full` correctness with hidden counts.
- Attendee-list AND-semantics: user opt-in + event hide ⇒ hidden; user opt-out + event show ⇒ still hidden.
- Recurring-event occurrence copy of `visibility_settings`.

### PR DESCRIPTION
**Pull request by biagiodistefano**
Closes #792. Frontend companion: letsrevel/revel-frontend#690 (blocked on this landing + `pnpm generate:api`).

## What ships

`Event.visibility_settings` — a JSONField validated by a pydantic `EventVisibilitySettings` model, so the FE gets a real typed object in OpenAPI rather than an opaque dict. Follows the existing `Event.schedule` JSONField-with-validator pattern (`events/utils/visibility_settings.py`, `extra="forbid"`, validated in `Event.clean()` so the admin raw-JSON widget is guarded too).

| toggle | type | default |
|---|---|---|
| `show_attendee_count` | bool | `true` |
| `show_capacity` | bool | `true` |
| `show_attendee_list` | bool | `true` |

Defaults reproduce today's behavior exactly; the stored representation for an untouched event is `{}` and every missing key reads as visible, so future toggles are backward-compatible for free. The migration is a plain `AddField` — no backfill, no existing event changes.

**Enforcement**: context-aware schema resolvers backed by cached model predicates, mirroring the existing `resolve_address` → `Event.can_user_see_address(user)` idiom. No controller-403s, no service-level filtering. Organization owners/staff and Django staff/superusers bypass every toggle (`Event.bypasses_visibility_settings`, instance-cached per user id).

**`is_full` is new and always public.** A derived boolean on `EventBaseSchema` (`effective_capacity > 0 and attendee_count >= effective_capacity`) so the FE keeps sold-out badges and blocked-RSVP copy when the numbers are gone. "Full yes/no" is not sensitive; exact numbers are. It reads the denormalized count and needs no extra query.

## Leak surface — all gated

- [x] `EventBaseSchema.attendee_count` — new `resolve_attendee_count`, `int | None`, gated by `show_attendee_count` (list + detail + calendar)
- [x] `EventBaseSchema.max_attendees` — new `resolve_max_attendees`, `int | None`, gated by `show_capacity`
- [x] `EventBaseSchema.seats_held` — `resolve_seats_held` gained a `context` argument and gates on `show_capacity` *before* the fallback COUNT query
- [x] `TicketTierSchema.total_available` — new `resolve_total_available`, gated via `tier.event.can_user_see_capacity(...)` (covers `GET /events/{id}/tickets/tiers` and every nested use in ticket/invitation schemas)
- [x] `EventPronounDistributionSchema` — totals are now `int | None` and, when counts are hidden for a non-privileged viewer, the service returns null totals **and an empty distribution**, without touching the DB. Redacting only the totals would have been theatre: the per-pronoun counts sum straight back to them.
- [x] Dietary summary per-restriction/preference `attendee_count` — now `int | None`, nulled for non-privileged viewers when counts are hidden. Entries and notes survive (you still need to know what to cook for); the numbers do not.
- [x] Attendee-list endpoint — `Event.attendees()` returns `none()` for a non-privileged viewer when `show_attendee_list` is off, **before** the `AttendeeVisibilityFlag` matrix. AND semantics: user opt-in + event hide ⇒ hidden; user opt-out + event show ⇒ still hidden.

Verified as not leaking today, untouched: total waitlist size is never public; waitlist position and `pending_offers_count` are self-only via `/my-status`.

## Known ceiling — documented, not fixed

**Seated ticketed events still reveal occupancy through the seat picker.** `GET /events/{id}/seating/availability` returns an anonymous per-seat sold/held map, and a seat map that hides occupancy cannot function as a seat map. `show_capacity` gates tier `total_available` numbers; the seat map stays as-is. There is a `# ponytail:` note on `Event.can_user_see_capacity` recording the ceiling and the upgrade path (a separate "blind seating" mode with server-side assignment only — a product decision, not a gating one). **Organizers running a seated event must be told that `show_capacity` does not cover them.**

## One thing worth flagging

Two of the gated fields are reached through a path where the resolver is handed an **already-built schema instance** rather than an ORM object: ninja re-validates whatever a route returns against its declared response schema, so `events.service.guest.confirm_guest_action` (which builds `BatchCheckoutResponse` with `UserTicketSchema.from_orm(...)`) runs `resolve_total_available` against a `TicketTierSchema` on the second pass. That instance has no `.event`. This is the same hazard already documented on `_CancellationResult` in `controllers/event_public/tickets.py`, and it was previously invisible because every other resolver in that file has a schema default that silently absorbs the resulting `AttributeError`.

Handled explicitly rather than by accident:
- `resolve_total_available` detects the schema-instance case and passes the (already-gated) value through, with a regression test.
- `common.schema.viewer_from_context` resolves the request user from a resolver context and returns `AnonymousUser` when there is none — so a context-less `from_orm` in a service, Celery task or email render fails *closed* rather than raising. Used by all five new gating resolvers.

## Not in scope

- Migrating `address_visibility` / `public_pronoun_distribution` into `visibility_settings` — phase 2 (#793), breaking contract change.
- Backend-side visibility presets — the FE owns preset bundles.

## Recurrence / duplication

`visibility_settings` is not in `_EXCLUDED_FROM_COPY`, so it rides along to duplicates and materialized occurrences (covered by tests, including a guard test that fails if anyone adds it to the exclusion set). Added to `PROPAGATABLE_FIELDS` so template edits reach unmodified future occurrences, and to `TemplateEditSchema` + `EventEditSchema`/`EventCreateSchema` so all the editors round-trip it.

## Contract changes for the FE

`attendee_count`, `max_attendees`, `seats_held`, `TicketTierSchema.total_available`, the three pronoun totals and the dietary `attendee_count` are all nullable now. On `total_available` `null` was already "unlimited" and now also means "not disclosed" — clients tell them apart via `event.visibility_settings.show_capacity`, which is serialized on every event response with all keys always present.

## Tests

- `src/events/tests/test_controllers/test_event_public/test_visibility_settings.py` — wire contract for every gated field across anonymous / authenticated-nonmember / attendee / org-staff / owner, toggles on and off, plus list + calendar, `is_full` with hidden counts, organizer create round-trip, and 422 on an unknown toggle.
- `src/events/tests/test_models/test_event_visibility_settings.py` — parsed-flag defaults, partial blobs, `clean()` rejection of unknown keys and wrong types, the bypass and its cache, `is_full` edge cases (uncapped, overbooked), attendee-list AND semantics, and occurrence/duplicate copy.
- `src/common/tests/test_viewer_from_context.py` — the context helper never raises.

`make check` is green (ruff, mypy --strict, migration check, i18n, file length, task names).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01AtpbeLss5QghtRPoTRVYV3

---

## Fixed during review

Two real bugs surfaced in review, both in the edit path rather than the read path:

1. **Partial edits replaced the whole settings blob** (da662b3b). Pydantic propagates `exclude_unset` into nested models, so a `PUT` carrying only `{"visibility_settings": {"show_capacity": false}}` wrote that single key over the stored one — an organizer who had hidden the attendee count had it re-disclosed by an unrelated edit. Reproduced end-to-end, then fixed with `build_visibility_settings_update`, which merges sent toggles onto stored ones for both the event-edit and template-edit paths (the latter also propagating the *merged* value to occurrences rather than the partial one). This makes the nested object consistent with every sibling field's "omit ⇒ unchanged" contract; deliberately naming a toggle still sets it.

2. **A default-equivalent write detached recurring occurrences** (0ed6cbc2). `{}` and `{"show_capacity": true}` mean the same thing, but the occurrence diff compared raw dicts, so writing the latter over the former set `is_modified=True` — permanently excluding the occurrence from `FUTURE_UNMODIFIED` template propagation over a no-op edit. Because the frontend counterpart round-trips the whole settings object on every save, this would have detached *every* occurrence on its first edit of any field. `_field_changed` now compares parsed `EventVisibilitySettings` for that one field. Storage stays sparse on purpose: `{}` meaning "all defaults" is what keeps adding a fourth toggle a non-event for existing rows.

A third comment (about the helper's explicit-`null` test) was partly right and partly not: the two edit schemas genuinely disagree — `EventEditSchema` rejects a null with a 422, `TemplateEditSchema` accepts one and the helper must treat it as "no change" or it would write NULL into a NOT NULL column. The hand-rolled test stand-in was replaced with the two real schemas so both behaviours are pinned, and the helper docstring (which had overclaimed uniform rejection) was corrected.

### CODE REVIEW

**Comment by biagiodistefano** [comment]
@coderabbitai review

File: src/events/schema/event.py [hunk 1]
```diff
@@ -55,6 +66,7 @@ class EventEditSchema(CityEditMixin):
         None, description="Deadline for submitting invitation requests or questionnaires"
     )
     can_attend_without_login: bool = False
+    visibility_settings: EventVisibilitySettingsSchema = Field(default_factory=EventVisibilitySettingsSchema)
```
[the remark this answers is not in the document]
Comment (biagiodistefano) [reply src/events/schema/event.py:72]: Valid, and worse than "trivial" — thanks. Fixed in da662b3b.

I reproduced it end-to-end before acting. Given an event stored as `{"show_attendee_count": false}`, a `PUT` of `{"visibility_settings": {"show_capacity": false}}` persisted `{"show_capacity": false}` and `show_attendee_count` flipped back to `true`. So this isn't only a parse-time default question: pydantic *does* propagate `exclude_unset` into the nested model (the payload dumps to exactly `{"show_capacity": False}`), and it's `update_db_instance` writing that partial blob **over** the stored one that loses the other toggles. Either way the outcome is the one that matters — an unrelated edit silently re-discloses a count the organizer had hidden, which is the exact failure this feature exists to prevent.

I went with merge rather than the documentation-only option, because documenting leaves the fail-open path in place.

The framing that settled it: merging makes `visibility_settings` **consistent** with its siblings, not special. Every other field on `EventEditSchema`/`TemplateEditSchema` already means "omit ⇒ unchanged" under `exclude_unset`; the nested object was the only place where omitting something reset it. Merge just extends that same contract to sub-key granularity. Turning a toggle back on still works — `{"show_attendee_count": true}` merges over a stored `false` — it simply has to be deliberate rather than a side effect.

Implementation, kept minimal:

- `build_visibility_settings_update(stored, payload)` in `events/utils/visibility_settings.py` — pure, no DB, no payload-type coupling beyond `BaseModel`, so both services import it without an import cycle and without reaching for a private cross-module name. Returns `{}` when the field wasn't sent, so callers splat it unconditionally.
- `event_update_service.update_event` and `recurrence_service.update_template` both use it. The template path additionally feeds the **merged** value into `changed_data`, so occurrences receive the merged blob rather than the partial one — that was a second, quieter instance of the same bug.
- An explicit `null` is deliberately *not* swallowed: it falls through to model validation and 400s like any other non-nullable field, rather than being silently reinterpreted as "no change".

Also took your docstring point on both schemas (`recurring_event.py` note alongside the existing per-field caveats, plus a comment on the `EventEditSchema` field).

Tests added: partial edit no longer re-enables a hidden toggle; a partial edit can still deliberately re-enable one; plus unit coverage for the helper (absent field, explicit null, partial merge, sent-wins-over-stored, `None` stored, and that the stored dict isn't mutated in place).

File: src/events/service/event_update_service.py [hunk 2]
```diff
@@ -144,7 +145,14 @@ def update_event(
...
 
-    updated_event = update_db_instance(event, payload, exclude={"series_pass_links"})
+    # Merge partial visibility toggles onto the stored blob rather than replacing
+    # it: naming one toggle must not silently re-enable the ones left out.
+    updated_event = update_db_instance(
+        event,
+        payload,
+        exclude={"series_pass_links"},
+        **build_visibility_settings_update(event.visibility_settings, payload),
+    )
```
[the remark this answers is not in the document]
Comment (biagiodistefano) [reply src/events/service/event_update_service.py:179]: Valid — reproduced and fixed in 0ed6cbc2.

Confirmed against the real endpoint before changing anything: an occurrence stored with `{}`, sent `{"show_capacity": true}` via `PUT`, persisted `{'show_capacity': True}` and came back with `is_modified: True`. Nothing about its visibility changed.

What makes this worth fixing rather than shrugging at: it breaks an invariant `update_event` already documents ("A no-op PUT does NOT mark it modified"), and the consequence is sticky — `FUTURE_UNMODIFIED` propagation skips modified occurrences forever after. Combined with the frontend counterpart (letsrevel/revel-frontend#690), which round-trips the whole `visibility_settings` object on every save, *every* occurrence would have been detached from its template on the first edit of any field. That's a much bigger blast radius than the visibility feature itself.

Fix is the normalized comparison you suggested, scoped to that one field:

```python
def _field_changed(field: str, before: t.Any, after: t.Any) -> bool:
    if field == "visibility_settings":
        return validate_visibility_settings(before) != validate_visibility_settings(after)
    return bool(before != after)
```

Everything else keeps plain `!=`, per the existing comment about which types the schema actually exposes. I deliberately did **not** fix this by canonicalizing storage (normalizing the blob to all three keys on save): `{}` meaning "all defaults" is what makes adding a fourth toggle later a non-event for existing rows, and freezing today's defaults into every row would throw that away.

Tests added: unit coverage on `_field_changed` (default-equivalent equal, real change detected, other fields unaffected), plus two endpoint tests asserting an all-defaults write leaves `is_modified` False while `{"show_capacity": false}` still sets it True.

File: src/events/tests/test_models/test_event_visibility_settings.py [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+        visibility_settings: EventVisibilitySettings | None = None
+
+    def test_absent_field_yields_no_write(self) -> None:
+        assert build_visibility_settings_update({"show_capacity": False}, self._Payload()) == {}
+
+    def test_explicit_null_yields_no_write(self) -> None:
+        """Left to model validation to reject, like any other non-nullable field."""
+        payload = self._Payload.model_validate({"visibility_settings": None})
+
+        assert build_visibility_settings_update({}, payload) == {}
```
[the remark this answers is not in the document]
Comment (biagiodistefano) [reply src/events/tests/test_models/test_event_visibility_settings.py]: Right that the test was misleading — fixed in 0ed6cbc2, though not quite the way you proposed, because checking it turned up something the suggestion would have hidden.

The two edit schemas genuinely disagree, and I verified both:

- `EventEditSchema.model_validate({"visibility_settings": None})` → **ValidationError**. Non-nullable, as you say; a null never reaches the helper.
- `TemplateEditSchema.model_validate({"visibility_settings": None})` → **accepted**, yields `None`. It declares `| None = None` like every one of its siblings on that schema.

So the null path *is* reachable, just via the template-edit endpoint rather than the event-edit one. Making `_Payload` non-nullable and asserting `ValidationError` would have modelled `EventEditSchema` correctly while dropping all coverage of the branch that actually executes — and the `isinstance(sent, dict)` guard is load-bearing there: writing `None` through would violate the column's NOT NULL.

Rather than fix the stand-in, I deleted it and pointed the tests at the two real schemas, so they can't drift from what the endpoints declare:

- `test_absent_field_yields_no_write` — parametrized over both schemas.
- `test_event_edit_schema_rejects_an_explicit_null` — asserts the `ValidationError` you were after.
- `test_template_edit_schema_treats_an_explicit_null_as_no_change` — pins the reachable branch.
- The merge cases now build payloads with `EventEditSchema`.

Your finding also caught a real docstring bug: `build_visibility_settings_update` claimed an explicit null is "left to model validation to reject, like any other non-nullable field", which is only true for one of the two callers. Corrected to spell out both.
